roachtest: include link to testeng grafana in issue posts

annrpom opened the pull request:
This adds a link, populated with relevant cluster name and test timeframe, to the testeng grafana instance for failed roachtests.

Fixes: #105894
Release note: None

annrpom committed:
roachtest: refactor createPostRequest to include necessary params

This change refactors createPostRequest in two ways:
 - It adds two new time.Time parameters, setting up work for populating
   the testeng Grafana link.
 - Error handling by invoking t.Fatalf has been replaced with returning
   an `error`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/107391)
<!-- Reviewable:end -->

tbg commented on pkg/cmd/roachtest/github_test.go:
```diff
@@ -103,6 +105,27 @@ func TestShouldPost(t *testing.T) {
+	generateHelpCommand(github)(r)
+
+	// TODO (BEFORE MERGE): time.Now() in tests == no bueno. could do regex, but want something like utils.TimeNow() that cc has
+	echotest.Require(t, strings.Replace(r.String(), "\n", "", -1), filepath.Join("testdata", "help_command.txt"))
```
Why remove the newlines? Looks weird in the output, don't we want the actual output?

annrpom replied:
hello - one benefit of removing newlines is that comparing actual vs expected outputs on error becomes a bit easer in Goland; however, yes the downside is that it looks weird in the output~ i shall fix this then.

annrpom replied:
hi again - i just remembered: it seems like echotest stops reading from the "expected" output file once it reads a newline

so something like

> echo
> ----
> See: stuff here
> 
> See: more stuff here

cuts off after "See: stuff here"

tbg replied:
Have you tried running with `-rewrite`? There is a syntax that allows for multi-line. I think it's

```
echo
----
----
foo

bar
----
----
```

but better let `-rewrite` do it properly.

annrpom replied:
ah! thank you for pointing this out

tbg commented on pkg/cmd/roachtest/github_test.go:
```diff
@@ -103,6 +105,27 @@ func TestShouldPost(t *testing.T) {
+	r := &issues.Renderer{}
+	generateHelpCommand(github)(r)
+
+	// TODO (BEFORE MERGE): time.Now() in tests == no bueno. could do regex, but want something like utils.TimeNow() that cc has
```
The usual way I avoid this is to pass in the `now` timestamp into `generateHelpCommand`.

tbg reviewed (commented):
Looks good so far, I know you have some updates cooking, so going to wait for those.

annrpom commented:
going to revert these changes for now on this PR to focus on refactor commits & then this will pop up again (so it will look more like refactor -> refactor -> main work)

annrpom committed:
roachtest: include link to testeng grafana in issue posts

This adds a link, populated with relevant cluster name and test timeframe,
to the testeng grafana instance for failed roachtests.

Fixes: #105894
Release note: None

annrpom commented on pkg/cmd/roachtest/github.go:
```diff
@@ -46,6 +47,26 @@ func roachtestPrefix(p string) string {
+		)(renderer)
+		issues.HelpCommandAsLink(
+			"Grafana",
+			fmt.Sprintf("https://go.crdb.dev/p/roachfana/%s/%d/%d", clusterName, start.Unix()*1000, end.Unix()*1000),
```
for the whole Unix()*1000 code: it looks like the grafana needs nsec to be included in the link? 

my reasoning for doing this was:
- there is "no data" whenever i remove nsec (here is a link where the start and end times are just unix epoch seconds: https://go.crdb.dev/p/roachfana/silvano-cluster/1689957243/1689957733)
- i wasn't sure if doing something like:
```
startSec := start.Unix()*1000 + start.UnixNano()%1e+9
```

would have been overkill (feel free to let me know!)

smg260 replied:
`start.UnixMilli()`? 

`Unix()` provides seconds

annrpom replied:
is there a facepalm emoji? no way that i forgot milliseconds existed... thank you

smg260 commented on pkg/cmd/roachtest/github.go:
```diff
@@ -98,14 +119,18 @@ func (g *githubIssues) shouldPost(t test.Test) (bool, string) {
 }
 
 func (g *githubIssues) createPostRequest(
```
minor: why enumerate all parameters instead of just passing `(t Test, message string)`?

annrpom replied:
ah so the `start` and `end` times belong to `*testImpl`; thus, those had to be passed in.

- the other thought was to change the `test.Test` type to `*testImpl` since `createPostRequest` only gets used with arguments of type `*testImpl` currently.
    - however, the conclusion to this (after speaking with Tobias) is that `*testImpl` is a "grab bag of stuff" that would lead to a more difficult time testing.
    
as for the `name` and `spec` fields, i did this since the function itself didn't seem to specifically need a `Test` other than to get the `name` and the `spec` (since t.Fatalf is now replaced). i am open to discuss this!

i believe this is one more parameter in the function signature than i promised Tobias...

smg260 replied:
Make sense - I tend not to like functions with long arg lists.

smg260 reviewed (commented):
Thanks for doing this. Minor comments

tbg reviewed (approved):
Thank you!

annrpom commented:
TFTR! ('-')7

bors r=smg260, tbg
